### PR TITLE
expression: implement vectorized evaluation for `SubXXXAndYYY`

### PR DESCRIPTION
**Pull request by SilvaXiang**
### What problem does this PR solve? 
Use go template to generate vectorized  11 SUBTIME functions from #12176 

- builtinSubDatetimeAndDurationSig

- builtinSubDatetimeAndStringSig

- builtinSubDurationAndDurationSig

- builtinSubDurationAndStringSig

- builtinSubStringAndDurationSig

- builtinSubStringAndStringSig

- builtinSubDateAndDurationSig

- builtinSubDateAndStringSig

- builtinSubTimeDateTimeNullSig

- builtinSubTimeStringNullSig

- builtinSubTimeDurationNullSig

### What is changed and how it works?

    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDatetimeAndDurationSig-VecBuiltinFunc-4               3922            302150 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDatetimeAndDurationSig-NonVecBuiltinFunc-4            3393            361119 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDatetimeAndStringSig-VecBuiltinFunc-4                  624           1899306 ns/op           94497 B/op       5814 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDatetimeAndStringSig-NonVecBuiltinFunc-4               624           2202165 ns/op           94320 B/op       5814 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDurationAndDurationSig-VecBuiltinFunc-4             108002             10695 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDurationAndDurationSig-NonVecBuiltinFunc-4           30878             38418 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDurationAndStringSig-VecBuiltinFunc-4                  768           1525616 ns/op           91830 B/op       5634 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDurationAndStringSig-NonVecBuiltinFunc-4               765           1602095 ns/op           91497 B/op       5634 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubStringAndDurationSig-VecBuiltinFunc-4                  510           2658512 ns/op          185895 B/op       9004 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubStringAndDurationSig-NonVecBuiltinFunc-4               465           2423924 ns/op          185699 B/op       9004 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubStringAndStringSig-VecBuiltinFunc-4                    382           3301695 ns/op          286293 B/op      14953 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubStringAndStringSig-NonVecBuiltinFunc-4                 367           3186760 ns/op          285774 B/op      14953 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDateAndDurationSig-VecBuiltinFunc-4                   2694            411262 ns/op           96050 B/op       3939 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDateAndDurationSig-NonVecBuiltinFunc-4                2307            477730 ns/op           96051 B/op       3939 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDateAndStringSig-VecBuiltinFunc-4                      544           2288155 ns/op          202358 B/op      10537 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubDateAndStringSig-NonVecBuiltinFunc-4                   484           2272191 ns/op          201554 B/op      10537 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubTimeDateTimeNullSig-VecBuiltinFunc-4              13504646                83.4 ns/op             0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubTimeDateTimeNullSig-NonVecBuiltinFunc-4              76623             15796 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubTimeStringNullSig-VecBuiltinFunc-4                  368790              3064 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubTimeStringNullSig-NonVecBuiltinFunc-4               147936              7918 ns/op               0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubTimeDurationNullSig-VecBuiltinFunc-4              14170660                90.3 ns/op             0 B/op          0 allocs/op
    BenchmarkVectorizedBuiltinTimeFuncGenerated/builtinSubTimeDurationNullSig-NonVecBuiltinFunc-4             111211              9843 ns/op               0 B/op          0 allocs/op

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by SunRunAway** [comment]
@SilvaXiang builtinSubXXXAndYYYSig may be resolved by go generate, could you take a look at https://docs.google.com/document/d/17XdYo_EwYaWANqbjS3i3OVSz3oLlovU2fIo1bRx4YqA/edit#

**Comment by SilvaXiang** [comment]
> @SilvaXiang builtinSubXXXAndYYYSig may be resolved by go generate, could you take a look at https://docs.google.com/document/d/17XdYo_EwYaWANqbjS3i3OVSz3oLlovU2fIo1bRx4YqA/edit#

请问一定只能通过Go generate方式改吗？能不能一个一个往builtin_time_generated_vec.go里面加需要修改的函数？

**Comment by SunRunAway** [comment]
> > @SilvaXiang builtinSubXXXAndYYYSig may be resolved by go generate, could you take a look at https://docs.google.com/document/d/17XdYo_EwYaWANqbjS3i3OVSz3oLlovU2fIo1bRx4YqA/edit#
> 
> 请问一定只能通过Go generate方式改吗？能不能一个一个往builtin_time_generated_vec.go里面加需要修改的函数？

Hi, since builtinAddXXXAndYYYSig is already templated, we'd like to template builtinSubXXXAndYYYSig too, so that it's better for us to maintain in the future. 

builtin_time_generated_vec.go is auto-generated by expression/generator/time_vec.go, each time we run `go generate`, the file will be regenerated, so it should not be modified by hands.


You may want to take a look at expression/generator/time_vec.go, try to template `sub` logic with `add`.

**Comment by SilvaXiang** [comment]
> > > @SilvaXiang builtinSubXXXAndYYYSig may be resolved by go generate, could you take a look at https://docs.google.com/document/d/17XdYo_EwYaWANqbjS3i3OVSz3oLlovU2fIo1bRx4YqA/edit#
> > 
> > 
> > 请问一定只能通过Go generate方式改吗？能不能一个一个往builtin_time_generated_vec.go里面加需要修改的函数？
> 
> Hi, since builtinAddXXXAndYYYSig is already templated, we'd like to template builtinSubXXXAndYYYSig too, so that it's better for us to maintain in the future.
> 
> builtin_time_generated_vec.go is auto-generated by expression/generator/time_vec.go, each time we run `go generate`, the file will be regenerated, so it should not be modified by hands.
> 
> You may want to take a look at expression/generator/time_vec.go, try to template `sub` logic with `add`.

Hi，I have finished 9 functions by `go generate`. Waiting for your review. Thanks.

**Comment by SilvaXiang** [comment]
@SunRunAway Hey,I have finished 9 functions by `go generate`. Waiting for your review. Thanks.

**Comment by SilvaXiang** [comment]
@qw4990 Hey,I have finished 9 functions by go generate. Waiting for your review. Thanks.

**Comment by SunRunAway** [comment]
It seems that you've miss two functions (builtinSubStringAndDurationSig and builtinSubStringAndStringSig).

File: expression/generator/time_vec.go [hunk 1]
```diff
@@ -609,6 +840,10 @@ func generateDotGo(fileName string) error {
 	if err != nil {
 		return err
 	}
+	err = subTime.Execute(w, subTimeSigsTmpl)
```
Comment (SunRunAway) [inline expression/generator/time_vec.go]: I've noticed that the template of `subTime` is similar as `addTime`, could you unify them into one template?

We can define an template viriable to distinguish `add` or `sub`

```
{{ $addOrSub = slice .FuncName 0 3 }}
```

then use `$addOrSub` in every difference.

then exectute template like this.
```go
addOrSubTime.Execute(w, function{FuncName: "AddTime", Sigs: addDataSigsTmpl})
addOrSubTime.Execute(w, function{FuncName: "SubTime", Sigs: subDataSigsTmpl})
```

**Comment by SilvaXiang** [comment]
> builtinSubDatetimeAndStringSig

ok, I will add it.

File: expression/generator/time_vec.go [hunk 2]
```diff
@@ -249,6 +250,196 @@ func (b *{{.SigName}}) vectorized() bool {
...
+		sum, err := types.Duration{Duration: arg0, Fsp: fsp0}.Sub(arg1Duration)
+		if err != nil {
+			return err
+		}
+		output := sum.String()
+	{{ end }}
+
+		// commit result
+	{{ if .Output.Fixed }}
+		resultSlice[i] = time.Duration(output)
```
[the remark this answers is not in the document]
Comment (SilvaXiang) [reply expression/generator/time_vec.go]: 虽然time.Duration是int64类型的，但是go语言也是需要一次转换的....

[the remark this answers is not in the document]
Comment (SilvaXiang) [reply expression/generator/time_vec.go]: Sorry , it is my fault

**Comment by SunRunAway** [comment]
Please consider this comment https://github.com/pingcap/tidb/pull/13691#discussion_r354163318

**Comment by qw4990** [comment]
@SilvaXiang Friendly ping, please resolve conflicts.

**Comment by SilvaXiang** [comment]
> @SilvaXiang Friendly ping, please resolve conflicts.

Ok, I will fix it tonight.

**Comment by SilvaXiang** [comment]
Sorry, I went out on Saturday and Sunday... I will continue to focus this in this week

**Comment by qw4990** [comment]
@SilvaXiang Friendly ping, please solve conflicts :D

File: expression/generator/time_vec.go (hunk 1 above)
Comment (francis0407) [reply expression/generator/time_vec.go]: @SilvaXiang Please address this comment.

Comment (SilvaXiang) [reply expression/generator/time_vec.go]: > @SilvaXiang Please address this comment.

OK , I will fix it this afternoon.

File: expression/generator/time_vec.go [hunk 3]
```diff
@@ -183,49 +166,159 @@ func (b *{{.SigName}}) vecEval{{ .Output.TypeName }}(input *chunk.Chunk, result
...
 		}
-	{{ else if eq .SigName "builtinAddDurationAndStringSig" }}
+		{{ else }}
+		arg0Duration := types.Duration{Duration: arg0, Fsp: -1}
+		arg1Duration := types.Duration{Duration: arg1, Fsp: -1}
+		outputDuration, err := arg0Duration.Sub(arg1Duration)
+		if err != nil {
+			return err
+		}
+		output := outputDuration.Duration
```
Comment (SunRunAway) [inline expression/generator/time_vec.go]: You can implement `types.SubDuration` and use it here.

File: expression/generator/time_vec.go [hunk 4]
```diff
@@ -183,49 +166,159 @@ func (b *{{.SigName}}) vecEval{{ .Output.TypeName }}(input *chunk.Chunk, result
...
 			return err
 		}
-	{{ else if eq .SigName "builtinAddStringAndDurationSig" }}
+		{{ else }}
+		arg0Duration := types.Duration{Duration: arg0, Fsp: -1}
+		outputDuration, err := arg0Duration.Sub(arg1Duration)
+		if err != nil {
+			return err
+		}
+		output := outputDuration.Duration
```
Comment (SunRunAway) [inline expression/generator/time_vec.go]: ditto.

File: expression/generator/time_vec.go [hunk 5]
```diff
@@ -234,7 +327,7 @@ func (b *{{.SigName}}) vecEval{{ .Output.TypeName }}(input *chunk.Chunk, result
 
 		// commit result
 	{{ if .Output.Fixed }}
-		resultSlice[i] = output
+		resultSlice[i] = output
```
Comment (SunRunAway) [inline expression/generator/time_vec.go]: This change is not needed.

**Comment by SunRunAway** [comment]
Hi, @SilvaXiang 
Just some tiny comments, PTAL

**Comment by SilvaXiang** [comment]
> Hi, @SilvaXiang
> Just some tiny comments, PTAL

Ok,I will fix it tonight

**Comment by SunRunAway** [comment]
@qw4990 PTAL.

**Pull request merged**